File: typescript/src/voice/VoiceCallService.ts
/**
 * Point-to-point voice call service.
 *
 * State machine:
 *   Outgoing: Idle → Ringing → Connected → Ended | Failed
 *   Incoming: Ringing → Connected → Ended | Failed
 *
 * Signaling uses PacketType.VoiceSignaling (JSON).
 * Audio frames use PacketType.VoiceCall (binary VoiceFrame).
 *
 * SPDX-License-Identifier: MIT
 */

import { DEFAULT_TTL } from "../constants.js";
import { MeshPacket } from "../protocol/MeshPacket.js";
import { PacketType } from "../protocol/PacketType.js";
import { IMeshSender } from "../routing/IMeshSender.js";

// ──────────────────────────────────────────────────────────────────
// Constants
// ──────────────────────────────────────────────────────────────────

const VOICE_CALL_TIMEOUT_SECONDS = 30;
const VOICE_FRAME_PRIORITY = 64;
const VOICE_SIGNALING_PRIORITY = 32;

// ──────────────────────────────────────────────────────────────────
// Wire types (snake_case — matches canonical JSON format)
// ──────────────────────────────────────────────────────────────────

type VoiceSignalingKind = "offer" | "answer" | "hangup" | "cancel" | "timeout";

interface VoiceSignalingMessage {
  kind: VoiceSignalingKind;
  call_id: string;
  from_uhid: string;
  to_uhid: string;
  proposed_codecs?: string[];
  selected_codec?: string;
  sample_rate_hz?: number;
  reason?: string;
}

// ──────────────────────────────────────────────────────────────────
// Public types
// ──────────────────────────────────────────────────────────────────

export type VoiceCallState = "idle" | "ringing" | "connected" | "ended" | "failed";

export interface VoiceCallInfo {
  callId: string;
  remoteUhid: string;
  state: VoiceCallState;
  isOutgoing: boolean;
  startedAt: Date;
}

export interface VoiceFrameEvent {
  callId: string;
  fromUhid: string;
  sequence: number;
  timestampMs: number;
  isSilence: boolean;
  encodedPayload: Uint8Array;
}

// ──────────────────────────────────────────────────────────────────
// Service
// ──────────────────────────────────────────────────────────────────

export class VoiceCallService {
  private readonly calls = new Map<string, VoiceCallInfo>();
  private readonly timeouts = new Map<string, ReturnType<typeof setTimeout>>();

  // Event callbacks
  onIncomingCall?: (info: VoiceCallInfo) => void;
  onCallConnected?: (info: VoiceCallInfo) => void;
  onCallEnded?: (info: VoiceCallInfo) => void;
  onFrameReceived?: (event: VoiceFrameEvent) => void;

  constructor(private readonly sender: IMeshSender) {}

  // ──────────────── outbound signaling ────────────────────────────

  async sendOffer(
    toUhid: string,
    codecs: string[],
    sampleRateHz: number,
  ): Promise<string> {
    if (!toUhid) throw new Error("toUhid must not be empty");

    const callId = crypto.randomUUID();
    const info: VoiceCallInfo = {
      callId,
      remoteUhid: toUhid,
      state: "ringing",
      isOutgoing: true,
      startedAt: new Date(),
    };
    this.calls.set(callId, info);
    this.armTimeout(callId);

    await this.sendSignaling({
      kind: "offer",
      call_id: callId,
      from_uhid: this.sender.localUhid,
      to_uhid: toUhid,
      proposed_codecs: codecs,
      sample_rate_hz: sampleRateHz,
    }, toUhid);

    return callId;
  }

  async acceptCall(callId: string): Promise<void> {
    const info = this.calls.get(callId);
    if (!info) return;
    this.clearTimeout(callId);
    info.state = "connected";
    this.onCallConnected?.(info);

    await this.sendSignaling({
      kind: "answer",
      call_id: callId,
      from_uhid: this.sender.localUhid,
      to_uhid: info.remoteUhid,
    }, info.remoteUhid);
  }

  async hangUp(callId: string): Promise<void> {
    const info = this.calls.get(callId);
    if (!info) return;
    this.clearTimeout(callId);
    const previousState = info.state;
    info.state = "ended";
    this.onCallEnded?.(info);

    const kind: VoiceSignalingKind = previousState === "ringing" ? "cancel" : "hangup";
    await this.sendSignaling({
      kind,
      call_id: callId,
      from_uhid: this.sender.localUhid,
      to_uhid: info.remoteUhid,
    }, info.remoteUhid);
  }

  // ──────────────── audio frame sending ───────────────────────────

  async sendFrame(
    callId: string,
    encodedAudio: Uint8Array,
    isSilence: boolean,
  ): Promise<boolean> {
    const info = this.calls.get(callId);
    if (!info || info.state !== "connected") return false;

    const payload = encodeVoiceFrame(callId, encodedAudio, isSilence);

    const packet = new MeshPacket();
    packet.type = PacketType.VoiceCall;
    packet.sourceUhid = this.sender.localUhid;
    packet.destinationUhid = info.remoteUhid;
    packet.ttl = DEFAULT_TTL;
    packet.priority = VOICE_FRAME_PRIORITY;
    packet.payload = payload;

    return this.sender.send(packet, info.remoteUhid);
  }

  // ──────────────── inbound packet handling ───────────────────────

  async onPacket(packet: MeshPacket): Promise<void> {
    switch (packet.type) {
      case PacketType.VoiceSignaling:
        await this.handleSignaling(packet);
        break;
      case PacketType.VoiceCall:
        this.handleFrame(packet);
        break;
      default:
        break;
    }
  }

  // ──────────────── inspection ─────────────────────────────────────

  getCallInfo(callId: string): VoiceCallInfo | undefined {
    return this.calls.get(callId);
  }

  getActiveCalls(): VoiceCallInfo[] {
    return Array.from(this.calls.values()).filter(
      (c) => c.state === "ringing" || c.state === "connected",
    );
  }

  // ──────────────── private helpers ────────────────────────────────

  private async handleSignaling(packet: MeshPacket): Promise<void> {
    let msg: VoiceSignalingMessage;
    try {
      msg = JSON.parse(new TextDecoder().decode(packet.payload)) as VoiceSignalingMessage;
    } catch {
      return;
    }
    if (!msg.call_id || !msg.kind) return;

    const { kind, call_id: callId, from_uhid: fromUhid } = msg;

    switch (kind) {
      case "offer": {
        if (this.calls.has(callId)) return; // duplicate
        const info: VoiceCallInfo = {
          callId,
          remoteUhid: fromUhid,
          state: "ringing",
          isOutgoing: false,
          startedAt: new Date(),
        };
        this.calls.set(callId, info);
        this.armTimeout(callId);
        this.onIncomingCall?.(info);
        break;
      }
      case "answer": {
        const info = this.calls.get(callId);
        if (!info || info.state !== "ringing") break;
        this.clearTimeout(callId);
        info.state = "connected";
        this.onCallConnected?.(info);
        break;
      }
      case "hangup":
      case "cancel":
      case "timeout": {
        const info = this.calls.get(callId);
        if (!info) break;
        this.clearTimeout(callId);
        info.state = "ended";
        this.onCallEnded?.(info);
        break;
      }
      default:
        break;
    }
  }

  private handleFrame(packet: MeshPacket): void {
    const frame = decodeVoiceFrame(packet.payload);
    if (!frame) return;
    const info = this.calls.get(frame.callId);
    if (!info || info.state !== "connected") return;

    this.onFrameReceived?.({
      callId: frame.callId,
      fromUhid: packet.sourceUhid,
      sequence: frame.sequence,
      timestampMs: frame.timestampMs,
      isSilence: frame.isSilence,
      encodedPayload: frame.encodedPayload,
    });
  }

  private async sendSignaling(msg: VoiceSignalingMessage, toUhid: string): Promise<void> {
    const body = new TextEncoder().encode(JSON.stringify(msg));
    const packet = new MeshPacket();
    packet.type = PacketType.VoiceSignaling;
    packet.sourceUhid = this.sender.localUhid;
    packet.destinationUhid = toUhid;
    packet.ttl = DEFAULT_TTL;
    packet.priority = VOICE_SIGNALING_PRIORITY;
    packet.payload = body;
    await this.sender.send(packet, toUhid);
  }

  private armTimeout(callId: string): void {
    this.clearTimeout(callId);
    const handle = setTimeout(async () => {
      const info = this.calls.get(callId);
      if (!info || info.state !== "ringing") return;
      info.state = "failed";
      this.onCallEnded?.(info);
      await this.sendSignaling({
        kind: "timeout",
        call_id: callId,
        from_uhid: this.sender.localUhid,
        to_uhid: info.remoteUhid,
        reason: "no_answer",
      }, info.remoteUhid);
    }, VOICE_CALL_TIMEOUT_SECONDS * 1000);
    this.timeouts.set(callId, handle);
  }

  private clearTimeout(callId: string): void {
    const handle = this.timeouts.get(callId);
    if (handle !== undefined) {
      clearTimeout(handle);
      this.timeouts.delete(callId);
    }
  }
}

// ──────────────────────────────────────────────────────────────────
// Binary codec helpers
// ──────────────────────────────────────────────────────────────────

/**
 * Encode VoiceFrame binary payload:
 * [16] CallId (UUID RFC4122 big-endian)
 * [4]  Sequence (uint32 little-endian)  — incrementing counter from timestamp
 * [8]  TimestampMs (int64 little-endian)
 * [1]  IsSilence (0 or 1)
 * [N]  EncodedPayload
 */
let voiceFrameSeq = 0;

export function encodeVoiceFrame(
  callId: string,
  encodedPayload: Uint8Array,
  isSilence: boolean,
): Uint8Array {
  const buf = new Uint8Array(16 + 4 + 8 + 1 + encodedPayload.length);
  const dv = new DataView(buf.buffer);

  uuidToBytes(callId, buf, 0);
  dv.setUint32(16, voiceFrameSeq++ & 0xffffffff, true);
  dv.setBigInt64(20, BigInt(Date.now()), true);
  buf[28] = isSilence ? 1 : 0;
  buf.set(encodedPayload, 29);

  return buf;
}

export function decodeVoiceFrame(data: Uint8Array): {
  callId: string;
  sequence: number;
  timestampMs: number;
  isSilence: boolean;
  encodedPayload: Uint8Array;
} | null {
  if (data.length < 29) return null;
  const dv = new DataView(data.buffer, data.byteOffset, data.byteLength);

  const callId = bytesToUuid(data, 0);
  const sequence = dv.getUint32(16, true);
  const timestampMs = Number(dv.getBigInt64(20, true));
  const isSilence = data[28] !== 0;
  const encodedPayload = data.slice(29);

  return { callId, sequence, timestampMs, isSilence, encodedPayload };
}

// ──────────────────────────────────────────────────────────────────
// UUID helpers shared across this file
// ──────────────────────────────────────────────────────────────────

export function uuidToBytes(uuid: string, buf: Uint8Array, offset: number): void {
  const hex = uuid.replace(/-/g, "");
  for (let i = 0; i < 16; i++) {
    buf[offset + i] = parseInt(hex.substr(i * 2, 2), 16);
  }
}

export function bytesToUuid(buf: Uint8Array, offset: number): string {
  const hex = Array.from(buf.slice(offset, offset + 16))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

File: kotlin/src/main/kotlin/aethermesh/voice/VoiceCallService.kt
// SPDX-License-Identifier: MIT

package aethermesh.voice

import aethermesh.AetherMeshConstants
import aethermesh.protocol.MeshPacket
import aethermesh.protocol.PacketType
import aethermesh.routing.MeshSender
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

// ─────────────────────────────────────────────────────────────────────────────
// Domain types
// ─────────────────────────────────────────────────────────────────────────────

enum class VoiceCallState { Outgoing, Incoming, Connected, Ended, Failed }

data class VoiceCallSession(
    val id: UUID,
    val callerUhid: String,
    val calleeUhid: String,
    var state: VoiceCallState,
    val proposedCodecs: List<String> = emptyList(),
    var selectedCodec: String? = null,
    var sampleRateHz: Int = 16000
)

/**
 * One-to-one voice call service.
 *
 * Wire format — VoiceFrame binary payload:
 *   [16] CallId  (UUID RFC4122 big-endian)
 *   [4]  Sequence (UInt32 little-endian)
 *   [8]  TimestampMs (Int64 little-endian)
 *   [1]  IsSilence (0 or 1)
 *   [N]  EncodedPayload
 *
 * Signaling uses VoiceSignalingMessage JSON (snake_case).
 *
 * Priority: 64 for voice frames, 32 for signaling.
 */
class VoiceCallService(private val sender: MeshSender) {

    private val sessions = ConcurrentHashMap<UUID, VoiceCallSession>()
    @Volatile private var frameSequence: Int = 0

    /** Invoked when an incoming call offer arrives. */
    var onIncomingCall: ((VoiceCallSession) -> Unit)? = null

    /** Invoked when a call state changes (answered, hung up, failed). */
    var onCallStateChanged: ((VoiceCallSession) -> Unit)? = null

    /** Invoked when a voice frame arrives for an active call. */
    var onFrameReceived: ((UUID, ByteArray, Boolean) -> Unit)? = null

    // ─────────────────────────────────────────────────────────────────────
    // Outbound API
    // ─────────────────────────────────────────────────────────────────────

    /**
     * Initiate a call to [toUhid]. Returns the new call-id.
     */
    suspend fun sendOffer(toUhid: String, codecs: List<String>, sampleRateHz: Int): UUID {
        require(toUhid.isNotEmpty()) { "toUhid must not be empty" }
        val callId = UUID.randomUUID()
        val session = VoiceCallSession(
            id = callId,
            callerUhid = sender.localUhid,
            calleeUhid = toUhid,
            state = VoiceCallState.Outgoing,
            proposedCodecs = codecs,
            sampleRateHz = sampleRateHz
        )
        sessions[callId] = session

        val payload = encodeSignaling(
            kind = "offer",
            callId = callId,
            fromUhid = sender.localUhid,
            toUhid = toUhid,
            proposedCodecs = codecs,
            sampleRateHz = sampleRateHz
        )
        sender.send(signalingPacket(toUhid, payload), toUhid)
        return callId
    }

    /**
     * Accept an incoming call. Sends an "answer" signaling message.
     */
    suspend fun acceptCall(callId: UUID) {
        val session = sessions[callId] ?: return
        if (session.state != VoiceCallState.Incoming) return
        session.state = VoiceCallState.Connected
        onCallStateChanged?.invoke(session)

        val payload = encodeSignaling(
            kind = "answer",
            callId = callId,
            fromUhid = sender.localUhid,
            toUhid = session.callerUhid,
            selectedCodec = session.selectedCodec,
            sampleRateHz = session.sampleRateHz
        )
        sender.send(signalingPacket(session.callerUhid, payload), session.callerUhid)
    }

    /**
     * Hang up or cancel [callId]. Sends a "hangup" or "cancel" to the remote peer.
     */
    suspend fun hangUp(callId: UUID) {
        val session = sessions.remove(callId) ?: return
        val kind = if (session.state == VoiceCallState.Outgoing) "cancel" else "hangup"
        val remote = if (sender.localUhid == session.callerUhid) session.calleeUhid else session.callerUhid
        session.state = VoiceCallState.Ended
        onCallStateChanged?.invoke(session)

        val payload = encodeSignaling(
            kind = kind,
            callId = callId,
            fromUhid = sender.localUhid,
            toUhid = remote
        )
        sender.send(signalingPacket(remote, payload), remote)
    }

    /**
     * Send an encoded audio frame for an active call.
     */
    suspend fun sendFrame(callId: UUID, encodedAudio: ByteArray, isSilence: Boolean) {
        val session = sessions[callId]
        if (session == null || session.state != VoiceCallState.Connected) return
        val remote = if (sender.localUhid == session.callerUhid) session.calleeUhid else session.callerUhid

        val seq = frameSequence++.toUInt()
        val payload = encodeVoiceFrame(callId, seq, System.currentTimeMillis(), isSilence, encodedAudio)
        val packet = MeshPacket(
            type = PacketType.VoiceCall,
            sourceUhid = sender.localUhid,
            destinationUhid = remote,
            ttl = AetherMeshConstants.DEFAULT_TTL,
            priority = 64,
            payload = payload
        )
        sender.send(packet, remote)
    }

    // ─────────────────────────────────────────────────────────────────────
    // Inbound packet dispatcher
    // ─────────────────────────────────────────────────────────────────────

    suspend fun handlePacket(packet: MeshPacket) {
        when (packet.type) {
            PacketType.VoiceSignaling -> handleSignaling(packet)
            PacketType.VoiceCall -> handleFrame(packet)
            else -> {}
        }
    }

    // ─────────────────────────────────────────────────────────────────────
    // Private helpers
    // ─────────────────────────────────────────────────────────────────────

    private fun handleSignaling(packet: MeshPacket) {
        val json = String(packet.payload, Charsets.UTF_8)
        val kind = JsonReader.readString(json, "kind") ?: return
        val callIdStr = JsonReader.readString(json, "call_id") ?: return
        val callId = runCatching { UUID.fromString(callIdStr) }.getOrNull() ?: return
        val fromUhid = JsonReader.readString(json, "from_uhid") ?: return

        when (kind) {
            "offer" -> {
                val codecs = JsonReader.readStringArray(json, "proposed_codecs")
                val rate = JsonReader.readInt(json, "sample_rate_hz") ?: 16000
                val session = VoiceCallSession(
                    id = callId,
                    callerUhid = fromUhid,
                    calleeUhid = sender.localUhid,
                    state = VoiceCallState.Incoming,
                    proposedCodecs = codecs,
                    sampleRateHz = rate
                )
                sessions[callId] = session
                onIncomingCall?.invoke(session)
            }
            "answer" -> {
                val session = sessions[callId] ?: return
                session.state = VoiceCallState.Connected
                session.selectedCodec = JsonReader.readString(json, "selected_codec")
                onCallStateChanged?.invoke(session)
            }
            "hangup", "cancel", "timeout" -> {
                val session = sessions.remove(callId) ?: return
                session.state = VoiceCallState.Ended
                onCallStateChanged?.invoke(session)
            }
        }
    }

    private fun handleFrame(packet: MeshPacket) {
        val data = packet.payload
        if (data.size < 29) return  // 16 + 4 + 8 + 1 = minimum 29 bytes
        val buf = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)

        // UUID is big-endian (first 16 bytes)
        val uuidBuf = ByteBuffer.wrap(data, 0, 16).order(ByteOrder.BIG_ENDIAN)
        val callId = UUID(uuidBuf.long, uuidBuf.long)
        buf.position(16)

        @Suppress("UNUSED_VARIABLE") val seq = buf.int.toUInt()
        @Suppress("UNUSED_VARIABLE") val ts = buf.long
        val isSilence = buf.get() != 0.toByte()
        val encoded = if (buf.hasRemaining()) {
            val arr = ByteArray(buf.remaining())
            buf.get(arr)
            arr
        } else ByteArray(0)

        onFrameReceived?.invoke(callId, encoded, isSilence)
    }

    private fun signalingPacket(toUhid: String, payload: ByteArray) = MeshPacket(
        type = PacketType.VoiceSignaling,
        sourceUhid = sender.localUhid,
        destinationUhid = toUhid,
        ttl = AetherMeshConstants.DEFAULT_TTL,
        priority = 32,
        payload = payload
    )
}

// ─────────────────────────────────────────────────────────────────────────────
// Wire helpers
// ─────────────────────────────────────────────────────────────────────────────

internal fun encodeVoiceFrame(
    callId: UUID,
    sequence: UInt,
    timestampMs: Long,
    isSilence: Boolean,
    encodedPayload: ByteArray
): ByteArray {
    val buf = ByteBuffer.allocate(16 + 4 + 8 + 1 + encodedPayload.size)
    buf.order(ByteOrder.BIG_ENDIAN)
    buf.putLong(callId.mostSignificantBits)
    buf.putLong(callId.leastSignificantBits)
    buf.order(ByteOrder.LITTLE_ENDIAN)
    buf.putInt(sequence.toInt())
    buf.putLong(timestampMs)
    buf.put(if (isSilence) 1.toByte() else 0.toByte())
    buf.put(encodedPayload)
    return buf.array()
}

internal fun encodeSignaling(
    kind: String,
    callId: UUID,
    fromUhid: String,
    toUhid: String,
    proposedCodecs: List<String>? = null,
    selectedCodec: String? = null,
    sampleRateHz: Int? = null,
    reason: String? = null
): ByteArray {
    val sb = StringBuilder()
    sb.append('{')
    sb.append("\"kind\":\"").append(jsonEscape(kind)).append("\",")
    sb.append("\"call_id\":\"").append(callId).append("\",")
    sb.append("\"from_uhid\":\"").append(jsonEscape(fromUhid)).append("\",")
    sb.append("\"to_uhid\":\"").append(jsonEscape(toUhid)).append('"')
    if (proposedCodecs != null) {
        sb.append(",\"proposed_codecs\":[")
        proposedCodecs.forEachIndexed { i, c ->
            if (i > 0) sb.append(',')
            sb.append('"').append(jsonEscape(c)).append('"')
        }
        sb.append(']')
    }
    if (selectedCodec != null) sb.append(",\"selected_codec\":\"").append(jsonEscape(selectedCodec)).append('"')
    if (sampleRateHz != null) sb.append(",\"sample_rate_hz\":").append(sampleRateHz)
    if (reason != null) sb.append(",\"reason\":\"").append(jsonEscape(reason)).append('"')
    sb.append('}')
    return sb.toString().toByteArray(Charsets.UTF_8)
}

// ─────────────────────────────────────────────────────────────────────────────
// Shared JSON utilities (package-private)
// ─────────────────────────────────────────────────────────────────────────────

internal object JsonReader {
    fun readString(json: String, key: String): String? {
        val needle = "\"$key\":"
        val i = json.indexOf(needle)
        if (i < 0) return null
        var p = i + needle.length
        while (p < json.length && json[p].isWhitespace()) p++
        if (p >= json.length || json[p] != '"') return null
        p++
        val sb = StringBuilder()
        while (p < json.length) {
            val c = json[p]
            if (c == '\\' && p + 1 < json.length) {
                when (json[p + 1]) {
                    'n' -> sb.append('\n')
                    'r' -> sb.append('\r')
                    't' -> sb.append('\t')
                    '"' -> sb.append('"')
                    '\\' -> sb.append('\\')
                    else -> sb.append(json[p + 1])
                }
                p += 2
            } else if (c == '"') {
                return sb.toString()
            } else {
                sb.append(c)
                p++
            }
        }
        return null
    }

    fun readInt(json: String, key: String): Int? = readLong(json, key)?.toInt()

    fun readLong(json: String, key: String): Long? {
        val needle = "\"$key\":"
        val i = json.indexOf(needle)
        if (i < 0) return null
        var p = i + needle.length
        while (p < json.length && json[p].isWhitespace()) p++
        val start = p
        while (p < json.length && (json[p].isDigit() || json[p] == '-')) p++
        if (p == start) return null
        return json.substring(start, p).toLongOrNull()
    }

    fun readDouble(json: String, key: String): Double? {
        val needle = "\"$key\":"
        val i = json.indexOf(needle)
        if (i < 0) return null
        var p = i + needle.length
        while (p < json.length && json[p].isWhitespace()) p++
        val start = p
        while (p < json.length && (json[p].isDigit() || json[p] == '-' || json[p] == '.' || json[p] == 'E' || json[p] == 'e' || json[p] == '+')) p++
        if (p == start) return null
        return json.substring(start, p).toDoubleOrNull()
    }

    fun readBool(json: String, key: String): Boolean? {
        val needle = "\"$key\":"
        val i = json.indexOf(needle)
        if (i < 0) return null
        var p = i + needle.length
        while (p < json.length && json[p].isWhitespace()) p++
        if (p + 4 <= json.length && json.regionMatches(p, "true", 0, 4)) return true
        if (p + 5 <= json.length && json.regionMatches(p, "false", 0, 5)) return false
        return null
    }

    fun readStringArray(json: String, key: String): List<String> {
        val needle = "\"$key\":["
        val i = json.indexOf(needle)
        if (i < 0) return emptyList()
        val start = i + needle.length
        val end = json.indexOf(']', start)
        if (end <= start) return emptyList()
        val body = json.substring(start, end).trim()
        if (body.isEmpty()) return emptyList()
        val result = mutableListOf<String>()
        var p = 0
        while (p < body.length) {
            while (p < body.length && body[p] != '"') p++
            if (p >= body.length) break
            p++
            val sb = StringBuilder()
            while (p < body.length) {
                val c = body[p]
                if (c == '\\' && p + 1 < body.length) { sb.append(body[p + 1]); p += 2 }
                else if (c == '"') { p++; break }
                else { sb.append(c); p++ }
            }
            result.add(sb.toString())
            while (p < body.length && body[p] != '"' && body[p] != ']') p++
        }
        return result
    }
}

internal fun jsonEscape(s: String): String {
    val sb = StringBuilder()
    for (c in s) {
        when (c) {
            '\\' -> sb.append("\\\\")
            '"' -> sb.append("\\\"")
            '\n' -> sb.append("\\n")
            '\r' -> sb.append("\\r")
            '\t' -> sb.append("\\t")
            else -> sb.append(c)
        }
    }
    return sb.toString()
}
